rust/wal3/src/lib.rs

```diff
@@ -0,0 +1,184 @@
+    /// The offset field of a LogPosition is a strictly increasing timestamp.  It has no gaps and
+    /// spans [0, u64::MAX).
+    offset: u64,
+    /// The timestamp_us field of a LogPosition is a strictly increasing timestamp.  It has gaps
+    /// and corresponds to wallclock time.
+    timestamp_us: u64,
+}
+
+impl LogPosition {
+    pub fn new(offset: u64, timestamp_us: u64) -> Self {
+        LogPosition {
+            offset,
+            timestamp_us,
+        }
+    }
+
+    pub fn offset(&self) -> u64 {
+        self.offset
+    }
+
+    pub fn timestamp_us(&self) -> u64 {
+        self.timestamp_us
+    }
+
+    pub fn contains_offset(start: LogPosition, end: LogPosition, offset: u64) -> bool {
+        start.offset <= offset && offset < end.offset
+    }
+
+    pub fn contains_timestamp(start: LogPosition, end: LogPosition, timestamp: u64) -> bool {
+        start.offset <= timestamp && timestamp < end.offset
+    }
+}
+
+////////////////////////////////////////// ThrottleOptions /////////////////////////////////////////
+
+/// ThrottleOptions control admission to S3 and batch size/interval.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct ThrottleOptions {
+    /// The maximum number of bytes to batch.  Defaults to 8MB.
+    pub batch_size: usize,
+    /// The maximum number of microseconds to batch.  Defaults to 20ms.
+    pub batch_interval: usize,
+    /// The maximum number of operations per second to allow.  Defaults to 2_000.
+    pub throughput: usize,
+    /// The number of operations per second to reserve for backoff/retry.  Defaults to 1_500.
+    pub headroom: usize,
+    /// The maximum number of outstanding requests to allow.  Defaults to 100.
+    pub outstanding: usize,
+}
+
+impl Default for ThrottleOptions {
+    fn default() -> Self {
+        ThrottleOptions {
+            // Batch for at least 20ms.
+            batch_interval: 20_000,
+            // Set a batch size of 8MB.
+            batch_size: 8 * 1_000_000,
+            // Set a throughput that's approximately 5/7th the throughput of the throughput S3
+            // allows.  If we hit throttle errors at this throughput we have a case for support.
+            throughput: 2_000,
+            // How much headroom we have for retries.
+            headroom: 1_500,
+            // Allow up to 100 requests to be outstanding.
+            outstanding: 100,
+        }
+    }
+}
+
+impl From<ThrottleOptions> for ExponentialBackoff {
+    fn from(options: ThrottleOptions) -> Self {
+        ExponentialBackoff::new(options.throughput as f64, options.headroom as f64)
+    }
+}
+
+////////////////////////////////////////// SnapshotOptions /////////////////////////////////////////
+
+/// SnapshotOptions control the size of snapshots and manifests.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, serde::Deserialize, serde::Serialize)]
+pub struct SnapshotOptions {
+    /// The maximum number of outbound snapshot pointers to embed in a snapshot or manifest.
```

Comment:
question i may answer by reading but leaving to not forget - how big are snapshot pointers, would it make more sense to configure this in bytes or similar and divide? As a configurer, its hard to know what this corresponds to.

Reply:
Up to you, just thinking out loud about tradeoffs.